File: truezip-kernel/src/main/java/de/schlichtherle/truezip/kernel/FinalizeController.java
/*
 * Copyright (C) 2005-2012 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package de.schlichtherle.truezip.kernel;

import de.truezip.kernel.*;
import de.truezip.kernel.cio.*;
import de.truezip.kernel.io.DecoratingInputStream;
import de.truezip.kernel.io.DecoratingOutputStream;
import de.truezip.kernel.io.DecoratingSeekableChannel;
import de.truezip.kernel.util.BitField;
import edu.umd.cs.findbugs.annotations.DischargesObligation;
import java.io.Closeable;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.channels.SeekableByteChannel;
import java.util.logging.Level;
import java.util.logging.Logger;
import javax.annotation.CheckForNull;
import javax.annotation.WillCloseWhenClosed;
import javax.annotation.concurrent.NotThreadSafe;
import javax.annotation.concurrent.ThreadSafe;

/**
 * Finalizes unclosed resources returned by its decorated controller.
 * 
 * @author Christian Schlichtherle
 */
@ThreadSafe
final class FinalizeController
extends FsDecoratingController<FsModel, FsController<?>> {

    private static final Logger logger = Logger.getLogger(
            FinalizeController.class.getName(),
            FinalizeController.class.getName());

    private static final IOException OK = new IOException((Throwable) null);

    /**
     * Constructs a new file system finalize controller.
     *
     * @param controller the decorated file system controller.
     */
    FinalizeController(FsController<?> controller) {
        super(controller);
    }

    @Override
    public InputSocket<?> getInputSocket(
            final FsEntryName name,
            final BitField<FsAccessOption> options) {
        @NotThreadSafe
        final class Input extends DecoratingInputSocket<Entry> {
            Input() {
                super(controller.getInputSocket(name, options));
            }

            @Override
            public InputStream stream() throws IOException {
                return new FinalizeInputStream(getBoundSocket().stream());
            }

            @Override
            public SeekableByteChannel channel() throws IOException {
                return new FinalizeSeekableChannel(getBoundSocket().channel());
            }
        } // Input

        return new Input();
    }

    @Override
    @edu.umd.cs.findbugs.annotations.SuppressWarnings("NP_PARAMETER_MUST_BE_NONNULL_BUT_MARKED_AS_NULLABLE") // false positive
    public OutputSocket<?> getOutputSocket(
            final FsEntryName name,
            final BitField<FsAccessOption> options,
            final @CheckForNull Entry template) {
        @NotThreadSafe
        final class Output extends DecoratingOutputSocket<Entry> {
            Output() {
                super(controller.getOutputSocket(name, options, template));
            }

            @Override
            public OutputStream stream() throws IOException {
                return new FinalizeOutputStream(getBoundSocket().stream());
            }

            @Override
            public SeekableByteChannel channel() throws IOException {
                return new FinalizeSeekableChannel(getBoundSocket().channel());
            }
        } // Output

        return new Output();
    }

    static void finalize(   final Closeable closeable,
                            final @CheckForNull IOException close) {
        if (OK == close) {
            logger.log(Level.FINEST, "closeCleared");
        } else if (null != close) {
            logger.log(Level.FINER, "closeFailed", close);
        } else {
            try {
                closeable.close();
                logger.log(Level.FINE, "finalizeCleared");
            } catch (final ControlFlowException ex) {  // report and swallow
                logger.log(Level.WARNING, "finalizeFailed",
                        new AssertionError("Unexpected control flow exception!", ex));
            } catch (final Throwable ex) {              // report and swallow
                logger.log(Level.INFO, "finalizeFailed", ex);
            }
        }
    }

    private static final class FinalizeInputStream
    extends DecoratingInputStream {
        volatile IOException close; // accessed by finalizer thread!

        FinalizeInputStream(@WillCloseWhenClosed InputStream in) {
            super(in);
        }

        @Override
        @DischargesObligation
        public void close() throws IOException {
            try {
                in.close();
            } catch (final IOException ex) {
                throw close = ex;
            }
            close = OK;
        }

        @Override
        @SuppressWarnings("FinalizeDeclaration")
        @DischargesObligation
        protected void finalize() throws Throwable {
            try {
                finalize(in, close);
            } finally {
                super.finalize();
            }
        }
    } // FinalizeInputStream

    private static final class FinalizeOutputStream
    extends DecoratingOutputStream {
        volatile IOException close; // accessed by finalizer thread!

        FinalizeOutputStream(@WillCloseWhenClosed OutputStream out) {
            super(out);
        }

        @Override
        @DischargesObligation
        public void close() throws IOException {
            try {
                out.close();
            } catch (final IOException ex) {
                throw close = ex;
            }
            close = OK;
        }

        @Override
        @SuppressWarnings("FinalizeDeclaration")
        @DischargesObligation
        protected void finalize() throws Throwable {
            try {
                finalize(out, close);
            } finally {
                super.finalize();
            }
        }
    } // FinalizeOutputStream

    private static final class FinalizeSeekableChannel
    extends DecoratingSeekableChannel {
        volatile IOException close; // accessed by finalizer thread!

        FinalizeSeekableChannel(@WillCloseWhenClosed SeekableByteChannel channel) {
            super(channel);
        }

        @Override
        @DischargesObligation
        public void close() throws IOException {
            try {
                channel.close();
            } catch (final IOException ex) {
                throw close = ex;
            }
            close = OK;
        }

        @Override
        @SuppressWarnings("FinalizeDeclaration")
        @DischargesObligation
        protected void finalize() throws Throwable {
            try {
                finalize(channel, close);
            } finally {
                super.finalize();
            }
        }
    } // FinalizeSeekableChannel
}

File: truevfs-kernel/truevfs-kernel-impl/src/main/scala/de/schlichtherle/truevfs/kernel/FinalizeController.scala
/*
 * Copyright (C) 2005-2012 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package de.schlichtherle.truevfs.kernel

import de.schlichtherle.truevfs.kernel._
import java.io._
import java.nio.channels._
import java.util.logging._
import java.util.logging.Level._
import javax.annotation.concurrent._
import net.truevfs.kernel._
import net.truevfs.kernel.cio._
import net.truevfs.kernel.cio.Entry._;
import net.truevfs.kernel.io._
import net.truevfs.kernel.util._

/** Finalizes unclosed resources returned by its decorated controller.
  * 
  * @author Christian Schlichtherle
  */
@ThreadSafe
private trait FinalizeController extends FsController[FsModel] {
  import FinalizeController._

  abstract override def input(options: AccessOptions, name: FsEntryName) = {
    final class Input extends DelegatingInputSocket[Entry] {
      val socket = FinalizeController.super.input(options, name)

      override def stream(peer: AnyOutputSocket) =
        new FinalizeInputStream(socket.stream(peer))

      override def channel(peer: AnyOutputSocket) =
        new FinalizeSeekableChannel(socket.channel(peer))
    }
    new Input
  }: AnyInputSocket

  abstract override def output(options: AccessOptions, name: FsEntryName, template: Entry) = {
    final class Output extends DelegatingOutputSocket[Entry] {
      val socket = FinalizeController.super.output(options, name, template)

      override def stream(peer: AnyInputSocket) =
        new FinalizeOutputStream(socket.stream(peer))

      override def channel(peer: AnyInputSocket) =
        new FinalizeSeekableChannel(socket.channel(peer))
    }
    new Output
  }: AnyOutputSocket
}

private object FinalizeController {
  private val logger = Logger.getLogger(
    classOf[FinalizeController].getName,
    classOf[FinalizeController].getName);

  private trait FinalizeResource extends Closeable {
    @volatile var ioException: Option[IOException] = _ // accessed by finalizer thread!

    abstract override def close() {
      try {
        super.close()
        ioException = None
      } catch {
        case ex: IOException => ioException = Some(ex); throw ex
      }
    }

    abstract override def finalize() {
      try {
        ioException match {
          case Some(ex) => logger.log(FINER, "closeFailed", ex);
          case None => logger.log(FINEST, "closeCleared");
          case _ =>
            try {
              super.close();
              logger.log(FINE, "finalizeCleared");
            } catch {
              case ex: ControlFlowException => // report and swallow
                logger.log(WARNING, "finalizeFailed",
                           new AssertionError("Unexpected control flow exception!", ex));
              case ex: Throwable => // report and swallow
                logger.log(INFO, "finalizeFailed", ex);
            }
        }
      } finally {
        super.finalize()
      }
    }
  } // FinalizeResource

  private final class FinalizeInputStream(in: InputStream)
  extends DecoratingInputStream(in) with FinalizeResource

  private final class FinalizeOutputStream(out: OutputStream)
  extends DecoratingOutputStream(out) with FinalizeResource

  private final class FinalizeSeekableChannel(channel: SeekableByteChannel)
  extends DecoratingSeekableChannel(channel) with FinalizeResource
}
